### PR TITLE
Contribution guidelines: type checks in lists

### PR DESCRIPTION
**Pull request by Rohesie**
* Updates the contribution guidelines to include the `as anything` key to skip checks in lists.
* Reduces the explanation a bit, to make it more succinct.
* Makes the examples use default byond types and vars, and code that actually compiles, instead of pseudo-code. Plus early continues, because they're good.

### CODE REVIEW

**Comment by Rohesie** [comment]
@tgstation/commit-access

**Comment by optimumtact** [comment]
gross

**Comment by Rohesie** [comment]
This is byond syntax, not really my choice.
I'm skeptical a define would make things much more clear, honestly.

**Comment by stylemistake** [comment]
```
#define of as anything
```

Nice and evil.

**Comment by Jared-Fogle** [comment]
+1 to `of` as a define, `as anything` looks weirdly relies-on-implementation-details to me.

**Comment by optimumtact** [comment]
-1 to the define, it's ugly but just accept it

**Comment by Cyberboss** [comment]
> 
> 
> -1 to the define, it's ugly but just accept it

I'd easily accept it if we could lint and auto-correct (i.e. prettier) to it.

**Comment by LemonInTheDark** [comment]
Is this more performant then 
```
for(var/item in list)
        var/typepath/varname = item
```
If so, by how much? I really really hate this syntax

**Comment by Rohesie** [comment]
I'll do whatever is decided, but I feel that an `of` define does not really offer any clarity as to what's going on, while `as anything`, ugly as it may be, at least has a parallel in filtering input (technically `as null|anything` would also work).

We don't have defines for filtering inputs, for good reasons.
It's ugly, sure, but it's DM syntax and I haven't seen a way to actually make it more clear yet.

> 
> 
> Is this more performant then
> 
> ```
> for(var/item in list)
>         var/typepath/varname = item
> ```
> 
> If so, by how much? I really really hate this syntax

By very little, performance shouldn't be a big concern here.
500000 iterations of a list containing 50 elements:
```
with the typecheck took 7265.44 ms
untyped with type casting took 6637.92 ms
as_anything took 5268.26 ms
```
But there's less boilerplate code involved.
It may be ugly, but I think we'll get used to it fast.

**Comment by stylemistake** [comment]
I was not suggesting that define seriously. You might remember, that I've also coined refactoring of `1 SECONDS` to `SECONDS(1)`.

**Comment by Rohesie** [comment]
> 
> 
> I was not suggesting that define seriously. You might remember, that I've also coined refactoring of `1 SECONDS` to `SECONDS(1)`.

That seconds define has the unironic support of every maintainer that has made their opinion public so far, though.
It just takes someone to do the refactor.

**Comment by Cyberboss** [comment]
Can at least add a grep lint for this? Something like: `for\s*\(\s*var\/[a-zA-Z_][a-zA-Z0-9_]*\s+(?!as\s+anything).+`

**Comment by Cyberboss** [comment]
Bit of an oof to setup
image

**Comment by Rohesie** [comment]
> 
> 
> Can we at least add a grep lint for this? Something like: `for\s*\(\s*var\/[a-zA-Z_][a-zA-Z0-9_]*\s+(?!as\s+anything\s+)in\s+.+\)`

To what end?
This catches things like `for(var/i in min(x1, x2) to max(x1, x2))` which is fine, as numbers don't have types.
Plus things like this that are not in immediate need of being typed (though it would add clarity if they were):
```dm
for(var/R in D.make_reagents)
	beaker.reagents.add_reagent(R, D.make_reagents[R])
```

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -355,6 +355,33 @@ This is good:
 	getter_turned_into_variable = condition ? VALUE_C : VALUE_D
 ```
 
+### Avoid unnecessary type checks and obscuring nulls in lists
+Typecasting in `for` loops carries and implied `istype()` check that filters non-matching types, nulls included. The `as anything` key can be used to skip the check.
```
Comment (MrDoomBringer) [inline .github/CONTRIBUTING.md:359]:
```suggestion
Typecasting in `for` loops carries an implied `istype()` check that filters non-matching types, nulls included. The `as anything` key can be used to skip the check.
```

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -355,6 +355,33 @@ This is good:
...
+
+If we know the list is supposed to only contain the desired type then we want to skip the check not only for the small optimization it offers, but also to catch any null entries that may creep into the list.
+
+Nulls in lists tend to point to improperly-handled references, which cause hard deletes are are hard to debug. Generating a runtime on those cases is more often than not positive.
```
Comment (MrDoomBringer) [inline .github/CONTRIBUTING.md:363]:
```suggestion
Nulls in lists tend to point to improperly-handled references, making hard deletes hard to debug. Generating a runtime in those cases is more often than not positive.
```

**Review by MrDoomBringer — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
> By very little, performance shouldn't be a big concern here.

If this is purely a code cleanliness thing I'm not in favor of standardizing it. I think we just disagree here, but just because it's a byond feature that doesn't mean we should use it. This looks like a straight downgrade in terms of readability to me.

**Comment by spookydonut** [comment]
> 
> 
> > By very little, performance shouldn't be a big concern here.
> 
> If this is purely a code cleanliness thing I'm not in favor of standardizing it. I think we just disagree here, but just because it's a byond feature that doesn't mean we should use it. This looks like a straight downgrade in terms of readability to me.

using this syntax over former typeless loop syntax removes two byte code operations per loop iteration, get+set.

according to zewaka it can have a huge performance impact.

**Comment by optimumtact** [comment]
according to zewaka eh

**Comment by LemonInTheDark** [comment]
Is it a cleanliness thing then? or is it a performance meme. If we're gonna standardize this we should know why

**Comment by Cyberboss** [comment]
I don't doubt we might get some small gain by replacing all ~4k fors with this. It does make readability a PITA though, which is why I still honestly support the `OF` define.

**Comment by Rohesie** [comment]
To me it's a cleanliness thing
```dm
for(var/obj/item/radio/radio as anything in radios)
	radio.stuff()
```
is much cleaner than
```dm
for(var/r in radios)
	var/obj/item/radio/radio = r
	radio.stuff()
```
Less boilerplate code overall, and `as flag` being something we already use elsewhere in the code.
We don't have issues with things like this, right?
`var/crystals = input("Amount of telecrystals for [key]","Syndicate uplink", U.telecrystals) as null | num`

**Comment by LemonInTheDark** [comment]
I personally find that syntax very confusing, even for input(), and I think it hurts understanding at a glance or hell even after explanation. Lines of code != complexity. Even with minor performance benefits I don't support this

**Comment by stylemistake** [comment]
You will get used to it.

**Comment by LemonInTheDark** [comment]
My concern is the people learning our codebase, I want the process of reading our code to be as painless as humanly possible, and adding a byondism to for loops is just gonna confuse people, even if it's possible to get used to.

**Comment by optimumtact** [comment]
suck it up buttercup

**Comment by LemonInTheDark** [comment]
You're right, it's depressing, but you're right

**Comment by Cyberboss** [comment]
It is what it is then.

**Pull request merged**